<a href="https://colab.research.google.com/github/AXM78251/Python_ML/blob/main/Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Classification**
***
Whereas linear regression was used to predict a numeric value (sych as the probability that a person x on the titanic will survive given various features), classification is used to separate data points into classes of different labels 

# **Imports & Setups**
***

In [1]:
%tensorflow_version 2.x  # this line is not required unless you are in a notebook

Colab only includes TensorFlow 2.x; %tensorflow_version has no effect.


In [2]:
from __future__ import absolute_import, division, print_function, unicode_literals


import tensorflow as tf

import pandas as pd

# **Dataset**
***
The specific dataset we are working with is the iris dataset, it seperates flowers into 3 different classes of species:
- Setosa
- Versicolor
- Virginica

The information about each flower is the following:
- sepal length
- sepal width
- petal length
- petal width

Ultimately, we will create a model that given the information/features about each flower, will predict which class/label the flower belongs to<br><br>

Next step is to define our csv column names and species

In [3]:
CSV_COLUMN_NAMES = ['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth', 'Species']
SPECIES = ['Setosa', 'Versicolor', 'Virginica']
# Lets define some constants to help us later on

In [4]:
train_path = tf.keras.utils.get_file(
    "iris_training.csv", "https://storage.googleapis.com/download.tensorflow.org/data/iris_training.csv")
test_path = tf.keras.utils.get_file(
    "iris_test.csv", "https://storage.googleapis.com/download.tensorflow.org/data/iris_test.csv")

train = pd.read_csv(train_path, names=CSV_COLUMN_NAMES, header=0)
test = pd.read_csv(test_path, names=CSV_COLUMN_NAMES, header=0)
# Here we use keras (a module inside of TensorFlow) to grab our datasets and read them into a pandas dataframe

16384/573 [=========================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================] - 0s 0us/step


Now let us look at the data

In [5]:
train.head()

SepalLength  SepalWidth  PetalLength  PetalWidth  Species
0          6.4         2.8          5.6         2.2        2
1          5.0         2.3          3.3         1.0        1
2          4.9         2.5          4.5         1.7        2
3          4.9         3.1          1.5         0.1        0
4          5.7         3.8          1.7         0.3        0

Now let us pop off the species column so that we may use it as our labels

In [6]:
train_y = train.pop('Species')
test_y = test.pop('Species')
train.head() # The species column is now gone

SepalLength  SepalWidth  PetalLength  PetalWidth
0          6.4         2.8          5.6         2.2
1          5.0         2.3          3.3         1.0
2          4.9         2.5          4.5         1.7
3          4.9         3.1          1.5         0.1
4          5.7         3.8          1.7         0.3

Next, print out the shape of our data

In [7]:
train.shape  # We have 120 entries with 4 features

(120, 4)

# **Input Function**
***
Similar to our linear regression model, we will need to implement an input function for our classification algorithm as well 


In [9]:
def input_fn(features, labels, training=True, batch_size=256):
    # Convert the inputs to a Dataset.
    dataset = tf.data.Dataset.from_tensor_slices((dict(features), labels))

    # Shuffle and repeat if you are in training mode.
    if training:
        dataset = dataset.shuffle(1000).repeat()
    
    return dataset.batch(batch_size)

# **Feature Columns**
***
Now, this is a tad easier since we are dealing with numerical values only

In [10]:
# Feature columns describe how to use the input.
my_feature_columns = []
for key in train.keys():  #Train.keys() will give us all the columns
    my_feature_columns.append(tf.feature_column.numeric_column(key=key))  #This will create/append our feature columns
print(my_feature_columns)

[NumericColumn(key='SepalLength', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None), NumericColumn(key='SepalWidth', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None), NumericColumn(key='PetalLength', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None), NumericColumn(key='PetalWidth', shape=(1,), default_value=None, dtype=tf.float32, normalizer_fn=None)]


#**Building our Model**
***
For classification tasks, ther are many estimators/models we can pick from, such as:


*   DNN Classifier (Deep Neural Network)
*   Linear Classifier

We will go with the DNN model because we may not be able to find a linear correspondence in our data<br><br>
So using this, we can now go ahead and build our model




In [11]:
# Build a DNN with 2 hidden layers with 30 and 10 hidden nodes each.
classifier = tf.estimator.DNNClassifier(
    feature_columns=my_feature_columns,
    # Two hidden layers of 30 and 10 nodes respectively.
    hidden_units=[30, 10],
    # The model must choose between 3 classes.
    n_classes=3)

INFO:tensorflow:Using default config.


INFO:tensorflow:Using default config.


INFO:tensorflow:Using config: {'_model_dir': '/tmp/tmpo28yz6nb', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_session_creation_timeout_secs': 7200, '_checkpoint_save_graph_def': True, '_service': None, '_cluster_spec': ClusterSpec({}), '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


INFO:tensorflow:Using config: {'_model_dir': '/tmp/tmpo28yz6nb', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_session_creation_timeout_secs': 7200, '_checkpoint_save_graph_def': True, '_service': None, '_cluster_spec': ClusterSpec({}), '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


What we've done above is create a deep neural network that has two hidden layers. These layers have 30 and 10 neurons respectively. This is the number of neurons the TensorFlow tutorial suggests. It is worth mentioning that the number of hidden neurons is an arbitrary number and many experiments and tests are usually done to determine the best choice for these values so it is possible to mess around with these values

# **Training our Model**
***
Now we can go ahead and train our model

In [12]:
classifier.train(
    input_fn=lambda: input_fn(train, train_y, training=True),
    steps=5000)
# We include a lambda to avoid creating an inner function previously

Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.


Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.


INFO:tensorflow:Calling model_fn.


INFO:tensorflow:Calling model_fn.


Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


INFO:tensorflow:Done calling model_fn.


INFO:tensorflow:Done calling model_fn.


INFO:tensorflow:Create CheckpointSaverHook.


INFO:tensorflow:Create CheckpointSaverHook.


INFO:tensorflow:Graph was finalized.


INFO:tensorflow:Graph was finalized.


INFO:tensorflow:Running local_init_op.


INFO:tensorflow:Running local_init_op.


INFO:tensorflow:Done running local_init_op.


INFO:tensorflow:Done running local_init_op.


INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 0...


INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 0...


INFO:tensorflow:Saving checkpoints for 0 into /tmp/tmpo28yz6nb/model.ckpt.


INFO:tensorflow:Saving checkpoints for 0 into /tmp/tmpo28yz6nb/model.ckpt.


INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 0...


INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 0...


INFO:tensorflow:loss = 1.516572, step = 0


INFO:tensorflow:loss = 1.516572, step = 0


INFO:tensorflow:global_step/sec: 387.652


INFO:tensorflow:global_step/sec: 387.652


INFO:tensorflow:loss = 1.1019311, step = 100 (0.266 sec)


INFO:tensorflow:loss = 1.1019311, step = 100 (0.266 sec)


INFO:tensorflow:global_step/sec: 474.513


INFO:tensorflow:global_step/sec: 474.513


INFO:tensorflow:loss = 1.0021954, step = 200 (0.209 sec)


INFO:tensorflow:loss = 1.0021954, step = 200 (0.209 sec)


INFO:tensorflow:global_step/sec: 454.39


INFO:tensorflow:global_step/sec: 454.39


INFO:tensorflow:loss = 0.95520747, step = 300 (0.223 sec)


INFO:tensorflow:loss = 0.95520747, step = 300 (0.223 sec)


INFO:tensorflow:global_step/sec: 464.032


INFO:tensorflow:global_step/sec: 464.032


INFO:tensorflow:loss = 0.92794585, step = 400 (0.215 sec)


INFO:tensorflow:loss = 0.92794585, step = 400 (0.215 sec)


INFO:tensorflow:global_step/sec: 474.548


INFO:tensorflow:global_step/sec: 474.548


INFO:tensorflow:loss = 0.89648, step = 500 (0.211 sec)


INFO:tensorflow:loss = 0.89648, step = 500 (0.211 sec)


INFO:tensorflow:global_step/sec: 466.684


INFO:tensorflow:global_step/sec: 466.684


INFO:tensorflow:loss = 0.8678412, step = 600 (0.213 sec)


INFO:tensorflow:loss = 0.8678412, step = 600 (0.213 sec)


INFO:tensorflow:global_step/sec: 454.33


INFO:tensorflow:global_step/sec: 454.33


INFO:tensorflow:loss = 0.8400936, step = 700 (0.221 sec)


INFO:tensorflow:loss = 0.8400936, step = 700 (0.221 sec)


INFO:tensorflow:global_step/sec: 428.119


INFO:tensorflow:global_step/sec: 428.119


INFO:tensorflow:loss = 0.81454515, step = 800 (0.233 sec)


INFO:tensorflow:loss = 0.81454515, step = 800 (0.233 sec)


INFO:tensorflow:global_step/sec: 462.765


INFO:tensorflow:global_step/sec: 462.765


INFO:tensorflow:loss = 0.79214185, step = 900 (0.216 sec)


INFO:tensorflow:loss = 0.79214185, step = 900 (0.216 sec)


INFO:tensorflow:global_step/sec: 477.458


INFO:tensorflow:global_step/sec: 477.458


INFO:tensorflow:loss = 0.77443796, step = 1000 (0.213 sec)


INFO:tensorflow:loss = 0.77443796, step = 1000 (0.213 sec)


INFO:tensorflow:global_step/sec: 458.076


INFO:tensorflow:global_step/sec: 458.076


INFO:tensorflow:loss = 0.7606058, step = 1100 (0.218 sec)


INFO:tensorflow:loss = 0.7606058, step = 1100 (0.218 sec)


INFO:tensorflow:global_step/sec: 461.309


INFO:tensorflow:global_step/sec: 461.309


INFO:tensorflow:loss = 0.73515433, step = 1200 (0.217 sec)


INFO:tensorflow:loss = 0.73515433, step = 1200 (0.217 sec)


INFO:tensorflow:global_step/sec: 401.865


INFO:tensorflow:global_step/sec: 401.865


INFO:tensorflow:loss = 0.7202041, step = 1300 (0.244 sec)


INFO:tensorflow:loss = 0.7202041, step = 1300 (0.244 sec)


INFO:tensorflow:global_step/sec: 467.187


INFO:tensorflow:global_step/sec: 467.187


INFO:tensorflow:loss = 0.70501834, step = 1400 (0.215 sec)


INFO:tensorflow:loss = 0.70501834, step = 1400 (0.215 sec)


INFO:tensorflow:global_step/sec: 474.594


INFO:tensorflow:global_step/sec: 474.594


INFO:tensorflow:loss = 0.6953517, step = 1500 (0.220 sec)


INFO:tensorflow:loss = 0.6953517, step = 1500 (0.220 sec)


INFO:tensorflow:global_step/sec: 453.9


INFO:tensorflow:global_step/sec: 453.9


INFO:tensorflow:loss = 0.6862139, step = 1600 (0.212 sec)


INFO:tensorflow:loss = 0.6862139, step = 1600 (0.212 sec)


INFO:tensorflow:global_step/sec: 451.63


INFO:tensorflow:global_step/sec: 451.63


INFO:tensorflow:loss = 0.66589415, step = 1700 (0.223 sec)


INFO:tensorflow:loss = 0.66589415, step = 1700 (0.223 sec)


INFO:tensorflow:global_step/sec: 457.134


INFO:tensorflow:global_step/sec: 457.134


INFO:tensorflow:loss = 0.65636754, step = 1800 (0.220 sec)


INFO:tensorflow:loss = 0.65636754, step = 1800 (0.220 sec)


INFO:tensorflow:global_step/sec: 473.428


INFO:tensorflow:global_step/sec: 473.428


INFO:tensorflow:loss = 0.63942283, step = 1900 (0.209 sec)


INFO:tensorflow:loss = 0.63942283, step = 1900 (0.209 sec)


INFO:tensorflow:global_step/sec: 463.204


INFO:tensorflow:global_step/sec: 463.204


INFO:tensorflow:loss = 0.62720096, step = 2000 (0.215 sec)


INFO:tensorflow:loss = 0.62720096, step = 2000 (0.215 sec)


INFO:tensorflow:global_step/sec: 468.956


INFO:tensorflow:global_step/sec: 468.956


INFO:tensorflow:loss = 0.6166927, step = 2100 (0.212 sec)


INFO:tensorflow:loss = 0.6166927, step = 2100 (0.212 sec)


INFO:tensorflow:global_step/sec: 449.307


INFO:tensorflow:global_step/sec: 449.307


INFO:tensorflow:loss = 0.61462986, step = 2200 (0.227 sec)


INFO:tensorflow:loss = 0.61462986, step = 2200 (0.227 sec)


INFO:tensorflow:global_step/sec: 452.586


INFO:tensorflow:global_step/sec: 452.586


INFO:tensorflow:loss = 0.59752774, step = 2300 (0.221 sec)


INFO:tensorflow:loss = 0.59752774, step = 2300 (0.221 sec)


INFO:tensorflow:global_step/sec: 458.505


INFO:tensorflow:global_step/sec: 458.505


INFO:tensorflow:loss = 0.58724594, step = 2400 (0.214 sec)


INFO:tensorflow:loss = 0.58724594, step = 2400 (0.214 sec)


INFO:tensorflow:global_step/sec: 477.516


INFO:tensorflow:global_step/sec: 477.516


INFO:tensorflow:loss = 0.5823548, step = 2500 (0.214 sec)


INFO:tensorflow:loss = 0.5823548, step = 2500 (0.214 sec)


INFO:tensorflow:global_step/sec: 470.641


INFO:tensorflow:global_step/sec: 470.641


INFO:tensorflow:loss = 0.57237375, step = 2600 (0.212 sec)


INFO:tensorflow:loss = 0.57237375, step = 2600 (0.212 sec)


INFO:tensorflow:global_step/sec: 457.495


INFO:tensorflow:global_step/sec: 457.495


INFO:tensorflow:loss = 0.5696064, step = 2700 (0.215 sec)


INFO:tensorflow:loss = 0.5696064, step = 2700 (0.215 sec)


INFO:tensorflow:global_step/sec: 502.002


INFO:tensorflow:global_step/sec: 502.002


INFO:tensorflow:loss = 0.54700613, step = 2800 (0.199 sec)


INFO:tensorflow:loss = 0.54700613, step = 2800 (0.199 sec)


INFO:tensorflow:global_step/sec: 492.112


INFO:tensorflow:global_step/sec: 492.112


INFO:tensorflow:loss = 0.55388737, step = 2900 (0.213 sec)


INFO:tensorflow:loss = 0.55388737, step = 2900 (0.213 sec)


INFO:tensorflow:global_step/sec: 454.742


INFO:tensorflow:global_step/sec: 454.742


INFO:tensorflow:loss = 0.5421822, step = 3000 (0.213 sec)


INFO:tensorflow:loss = 0.5421822, step = 3000 (0.213 sec)


INFO:tensorflow:global_step/sec: 480.834


INFO:tensorflow:global_step/sec: 480.834


INFO:tensorflow:loss = 0.5384345, step = 3100 (0.208 sec)


INFO:tensorflow:loss = 0.5384345, step = 3100 (0.208 sec)


INFO:tensorflow:global_step/sec: 462.269


INFO:tensorflow:global_step/sec: 462.269


INFO:tensorflow:loss = 0.5249014, step = 3200 (0.216 sec)


INFO:tensorflow:loss = 0.5249014, step = 3200 (0.216 sec)


INFO:tensorflow:global_step/sec: 466.052


INFO:tensorflow:global_step/sec: 466.052


INFO:tensorflow:loss = 0.52782035, step = 3300 (0.212 sec)


INFO:tensorflow:loss = 0.52782035, step = 3300 (0.212 sec)


INFO:tensorflow:global_step/sec: 494.481


INFO:tensorflow:global_step/sec: 494.481


INFO:tensorflow:loss = 0.5132315, step = 3400 (0.206 sec)


INFO:tensorflow:loss = 0.5132315, step = 3400 (0.206 sec)


INFO:tensorflow:global_step/sec: 486.788


INFO:tensorflow:global_step/sec: 486.788


INFO:tensorflow:loss = 0.50952935, step = 3500 (0.209 sec)


INFO:tensorflow:loss = 0.50952935, step = 3500 (0.209 sec)


INFO:tensorflow:global_step/sec: 442.668


INFO:tensorflow:global_step/sec: 442.668


INFO:tensorflow:loss = 0.51536345, step = 3600 (0.221 sec)


INFO:tensorflow:loss = 0.51536345, step = 3600 (0.221 sec)


INFO:tensorflow:global_step/sec: 477.205


INFO:tensorflow:global_step/sec: 477.205


INFO:tensorflow:loss = 0.49558592, step = 3700 (0.208 sec)


INFO:tensorflow:loss = 0.49558592, step = 3700 (0.208 sec)


INFO:tensorflow:global_step/sec: 480.331


INFO:tensorflow:global_step/sec: 480.331


INFO:tensorflow:loss = 0.49104694, step = 3800 (0.214 sec)


INFO:tensorflow:loss = 0.49104694, step = 3800 (0.214 sec)


INFO:tensorflow:global_step/sec: 474.881


INFO:tensorflow:global_step/sec: 474.881


INFO:tensorflow:loss = 0.49362206, step = 3900 (0.206 sec)


INFO:tensorflow:loss = 0.49362206, step = 3900 (0.206 sec)


INFO:tensorflow:global_step/sec: 501.733


INFO:tensorflow:global_step/sec: 501.733


INFO:tensorflow:loss = 0.50070155, step = 4000 (0.199 sec)


INFO:tensorflow:loss = 0.50070155, step = 4000 (0.199 sec)


INFO:tensorflow:global_step/sec: 437.347


INFO:tensorflow:global_step/sec: 437.347


INFO:tensorflow:loss = 0.47520277, step = 4100 (0.230 sec)


INFO:tensorflow:loss = 0.47520277, step = 4100 (0.230 sec)


INFO:tensorflow:global_step/sec: 476.133


INFO:tensorflow:global_step/sec: 476.133


INFO:tensorflow:loss = 0.4810393, step = 4200 (0.213 sec)


INFO:tensorflow:loss = 0.4810393, step = 4200 (0.213 sec)


INFO:tensorflow:global_step/sec: 481.83


INFO:tensorflow:global_step/sec: 481.83


INFO:tensorflow:loss = 0.47072193, step = 4300 (0.202 sec)


INFO:tensorflow:loss = 0.47072193, step = 4300 (0.202 sec)


INFO:tensorflow:global_step/sec: 478.596


INFO:tensorflow:global_step/sec: 478.596


INFO:tensorflow:loss = 0.47143573, step = 4400 (0.208 sec)


INFO:tensorflow:loss = 0.47143573, step = 4400 (0.208 sec)


INFO:tensorflow:global_step/sec: 479.608


INFO:tensorflow:global_step/sec: 479.608


INFO:tensorflow:loss = 0.46672142, step = 4500 (0.209 sec)


INFO:tensorflow:loss = 0.46672142, step = 4500 (0.209 sec)


INFO:tensorflow:global_step/sec: 441.106


INFO:tensorflow:global_step/sec: 441.106


INFO:tensorflow:loss = 0.469827, step = 4600 (0.226 sec)


INFO:tensorflow:loss = 0.469827, step = 4600 (0.226 sec)


INFO:tensorflow:global_step/sec: 471.246


INFO:tensorflow:global_step/sec: 471.246


INFO:tensorflow:loss = 0.4546725, step = 4700 (0.214 sec)


INFO:tensorflow:loss = 0.4546725, step = 4700 (0.214 sec)


INFO:tensorflow:global_step/sec: 436.118


INFO:tensorflow:global_step/sec: 436.118


INFO:tensorflow:loss = 0.45583752, step = 4800 (0.236 sec)


INFO:tensorflow:loss = 0.45583752, step = 4800 (0.236 sec)


INFO:tensorflow:global_step/sec: 426.859


INFO:tensorflow:global_step/sec: 426.859


INFO:tensorflow:loss = 0.4396662, step = 4900 (0.228 sec)


INFO:tensorflow:loss = 0.4396662, step = 4900 (0.228 sec)


INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 5000...


INFO:tensorflow:Calling checkpoint listeners before saving checkpoint 5000...


INFO:tensorflow:Saving checkpoints for 5000 into /tmp/tmpo28yz6nb/model.ckpt.


INFO:tensorflow:Saving checkpoints for 5000 into /tmp/tmpo28yz6nb/model.ckpt.


INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 5000...


INFO:tensorflow:Calling checkpoint listeners after saving checkpoint 5000...


INFO:tensorflow:Loss for final step: 0.43938422.


INFO:tensorflow:Loss for final step: 0.43938422.


In the code block above, we introduce the concept of lambdas which are 1 line functions where anything after the ":" is automatically returned without having to specify the return keyword<br><br>

After the code has finished executing, we see that the loss for the final step is 0.43938422 which is pretty high (and pretty bad), of course this is okay for now since it is our first attempt at a neural network but in the future, we want to make sure this va;ue is as low as possible<br><br>

As the model is trained, we see output being printed out to the console, since our model is small we dom't really care, but for bigger models that take terabytes of data (for example), then we definitely care about the progress of them 

# **Evaluation of our Model**
***
Now we can see how our trained model performs

In [13]:
eval_result = classifier.evaluate(
    input_fn=lambda: input_fn(test, test_y, training=False))

print('\nTest set accuracy: {accuracy:0.3f}\n'.format(**eval_result))

INFO:tensorflow:Calling model_fn.


INFO:tensorflow:Calling model_fn.


INFO:tensorflow:Done calling model_fn.


INFO:tensorflow:Done calling model_fn.


INFO:tensorflow:Starting evaluation at 2022-08-04T06:26:54


INFO:tensorflow:Starting evaluation at 2022-08-04T06:26:54


INFO:tensorflow:Graph was finalized.


INFO:tensorflow:Graph was finalized.


INFO:tensorflow:Restoring parameters from /tmp/tmpo28yz6nb/model.ckpt-5000


INFO:tensorflow:Restoring parameters from /tmp/tmpo28yz6nb/model.ckpt-5000


INFO:tensorflow:Running local_init_op.


INFO:tensorflow:Running local_init_op.


INFO:tensorflow:Done running local_init_op.


INFO:tensorflow:Done running local_init_op.


INFO:tensorflow:Inference Time : 0.58257s


INFO:tensorflow:Inference Time : 0.58257s


INFO:tensorflow:Finished evaluation at 2022-08-04-06:26:55


INFO:tensorflow:Finished evaluation at 2022-08-04-06:26:55


INFO:tensorflow:Saving dict for global step 5000: accuracy = 0.8666667, average_loss = 0.52023184, global_step = 5000, loss = 0.52023184


INFO:tensorflow:Saving dict for global step 5000: accuracy = 0.8666667, average_loss = 0.52023184, global_step = 5000, loss = 0.52023184


INFO:tensorflow:Saving 'checkpoint_path' summary for global step 5000: /tmp/tmpo28yz6nb/model.ckpt-5000


INFO:tensorflow:Saving 'checkpoint_path' summary for global step 5000: /tmp/tmpo28yz6nb/model.ckpt-5000



Test set accuracy: 0.867



This time we didn't specify the number of steps and this is because during evaluation of our model, we will only look at the testing data obe time

# **Predictions with our Model**
***
Now that we have a trained model, we can use it to make predictions for any given flower

In [16]:
def input_fn(features, batch_size=256):
    # Convert the inputs to a Dataset without labels.
    return tf.data.Dataset.from_tensor_slices(dict(features)).batch(batch_size)

features = ['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth']
predict = {}

print("Please type numeric values as prompted.")
# For each feature, wait to get a valid response
# Once we get some valid response, we'll add that to our dictionary 
for feature in features:
  valid = True
  while valid: 
    val = input(feature + ": ")
    if not val.isdigit(): valid = False

  predict[feature] = [float(val)]

predictions = classifier.predict(input_fn=lambda: input_fn(predict))
for pred_dict in predictions:
    print(pred_dict)
    class_id = pred_dict['class_ids'][0]  # The id of the class it predicts the flower is
    probability = pred_dict['probabilities'][class_id] # The probability of it being the class, where the class is specified by the class id

    print('Prediction is "{}" ({:.1f}%)'.format(
        SPECIES[class_id], 100 * probability))


Please type numeric values as prompted.
SepalLength: 1.4
SepalWidth: 2.3
PetalLength: 4.5
PetalWidth: 4.7
INFO:tensorflow:Calling model_fn.


INFO:tensorflow:Calling model_fn.


INFO:tensorflow:Done calling model_fn.


INFO:tensorflow:Done calling model_fn.


INFO:tensorflow:Graph was finalized.


INFO:tensorflow:Graph was finalized.


INFO:tensorflow:Restoring parameters from /tmp/tmpo28yz6nb/model.ckpt-5000


INFO:tensorflow:Restoring parameters from /tmp/tmpo28yz6nb/model.ckpt-5000


INFO:tensorflow:Running local_init_op.


INFO:tensorflow:Running local_init_op.


INFO:tensorflow:Done running local_init_op.


INFO:tensorflow:Done running local_init_op.


{'logits': array([-1.3872693,  3.409126 ,  4.118858 ], dtype=float32), 'probabilities': array([0.00271541, 0.32876295, 0.66852164], dtype=float32), 'class_ids': array([2]), 'classes': array([b'2'], dtype=object), 'all_class_ids': array([0, 1, 2], dtype=int32), 'all_classes': array([b'0', b'1', b'2'], dtype=object)}
Prediction is "Virginica" (66.9%)


The script above allows the user to type the features of a flower and see a prediction for its class

In [ ]:
# Here is some example input and expected classes you can try above
expected = ['Setosa', 'Versicolor', 'Virginica']
predict_x = {
    'SepalLength': [5.1, 5.9, 6.9],
    'SepalWidth': [3.3, 3.0, 3.1],
    'PetalLength': [1.7, 4.2, 5.4],
    'PetalWidth': [0.5, 1.5, 2.1],
}

The above are just examples we can try out on our own<br><br>
This is the end of our classification implementation uasing TensorFlow

All credit for code/material above goes to Tim Ruscica, otherwise known on youtube as "Tech With Tim" <br>
Youtube: ["Tech With Tim"](https://www.youtube.com/c/TechWithTim)